### PR TITLE
feature: Dashboard filters

### LINKED ISSUE #7113: Global dashboard filters

### PR DESCRIPTION
**Pull request by danharrin**
Closes #7113.

### CODE REVIEW

File: packages/support/docs/03-icons.md [hunk 1]
```diff
@@ -69,12 +69,13 @@ Alternatively, you may pass an SVG element into the component's slot instead of
 ### Panel Builder icon aliases
 
 - `panels::global-search.field` - Global search field
+- `panels::pages.dashboard.actions.filter.grouped` - Trigger button of the dashboard filter action
```
Comment (zepfietje) [inline packages/support/docs/03-icons.md]:
```suggestion
- `panels::pages.dashboard.actions.filter.grouped` - Trigger button of the grouped dashboard filter action
```

File: packages/panels/src/Pages/Dashboard/Actions/FilterAction.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        parent::setUp();
+
+        $this->label(__('filament-panels::pages/dashboard.actions.filter.label'));
+
+        $this->modalHeading(fn (): string => __('filament-panels::pages/dashboard.actions.filter.modal.heading'));
+
+        $this->modalSubmitActionLabel(__('filament-panels::pages/dashboard.actions.filter.modal.actions.apply.label'));
+
+        $this->icon(FilamentIcon::resolve('panels::pages.dashboard.actions.filter.grouped') ?? 'heroicon-m-funnel');
```
Comment (zepfietje) [inline packages/panels/src/Pages/Dashboard/Actions/FilterAction.php]: Should this be grouped, or should the alias be updated?

File: packages/panels/resources/views/pages/dashboard.blade.php [hunk 3]
```diff
@@ -1,7 +1,18 @@
...
+            {{ $this->filtersForm }}
+        </x-filament::section>
+    @endif
+
     <x-filament-widgets::widgets
         :columns="$this->getColumns()"
-        :data="$this->getWidgetData()"
+        :data="
+            [
+                ...(property_exists($this, 'filters') ? ['filters' => $this->filters] : []),
```
Comment (zepfietje) [inline packages/panels/resources/views/pages/dashboard.blade.php:12]: Isn't it cleaner to check for usage of the trait?

File: packages/panels/docs/05-dashboard.md [hunk 4]
```diff
@@ -109,6 +109,107 @@ php artisan make:filament-widget BlogPostsOverview
 
 This command will create two files - a widget class in the `/Widgets` directory of the Filament directory, and a view in the `/widgets` directory of the Filament views directory.
 
+## Filtering widget data
+
+You may add a form to the dashboard that allows the user to filter the data displayed across all widgets. When the filters are updated, the widgets will be reloaded with the new data.
+
+Firstly, you must [replace the original Dashboard page](#customizing-the-dashboard-page).
+
+Now, in your new `app/Filament/Pages/Dashboard.php` file, you may add the `HasFiltersForm` trait, and add the `filtersForm()` method to return add form components:
```
Comment (zepfietje) [inline packages/panels/docs/05-dashboard.md]:
```suggestion
Now, in your new `app/Filament/Pages/Dashboard.php` file, you may add the `HasFiltersForm` trait, and add the `filtersForm()` method to return form components:
```

File: packages/panels/resources/views/pages/dashboard.blade.php [hunk 5]
```diff
@@ -1,7 +1,18 @@
 <x-filament-panels::page class="fi-dashboard-page">
+    @if (method_exists($this, 'filtersForm'))
+        <x-filament::section>
```
Comment (zepfietje) [inline packages/panels/resources/views/pages/dashboard.blade.php:3]: Maybe we should add a heading to this section?

**Review by zepfietje — CHANGES_REQUESTED** [review]

File: packages/panels/resources/views/pages/dashboard.blade.php (hunk 5 above)
Comment (danharrin) [reply packages/panels/resources/views/pages/dashboard.blade.php:3]: I tried, but IMO it takes up too much vertical space

Comment (zepfietje) [reply packages/panels/resources/views/pages/dashboard.blade.php:3]: Alright.

File: packages/panels/src/Pages/Dashboard/Actions/FilterAction.php (hunk 2 above)
Comment (zepfietje) [reply packages/panels/src/Pages/Dashboard/Actions/FilterAction.php]: @danharrin

File: packages/support/docs/03-icons.md (hunk 1 above)
Comment (zepfietje) [reply packages/support/docs/03-icons.md]: @danharrin

File: packages/panels/resources/views/pages/dashboard.blade.php (hunk 3 above)
Comment (zepfietje) [reply packages/panels/resources/views/pages/dashboard.blade.php:12]: @danharrin

Comment (danharrin) [reply packages/panels/resources/views/pages/dashboard.blade.php:12]: This is faster, less code (no import), and in reality we don't need any features of the trait, just the property to exist so we can pass it in to widgets. I don't think it hurts

File: packages/panels/src/Pages/Dashboard/Actions/FilterAction.php (hunk 2 above)
Comment (danharrin) [reply packages/panels/src/Pages/Dashboard/Actions/FilterAction.php]: Updated the alias to remove grouped

File: packages/support/docs/03-icons.md (hunk 1 above)
Comment (danharrin) [reply packages/support/docs/03-icons.md]: Updated the alias to remove grouped

**Comment by patriktoth67** [comment]
Hi guys, if on my widget I do:
```
protected static ?string $pollingInterval = null;
```
or 
```
protected static ?string $pollingInterval = '10s';
```
the widgets don't update on filter change anymore. Possibly a conflict with Polling Interval's logic and this.

**Comment by patriktoth67** [comment]
Hi guys. This feature works perfectly on BaseWidget-s, but sadly doesn't work properly with ChartWidget-s.
If you try to use it on a chart, you can see the values changing in the livewire component, but not in chart.js sadly. It can either update if the pollingInterval is not null, or if you refresh the page. If pollingInterval is 5s, it only updates with the correct filtered data once the 5s is up.

**Comment by danharrin** [comment]
Please open an issue with a reproduction repository

**Comment by danielreales7** [comment]
This part of documentation https://filamentphp.com/docs/3.x/panels/dashboard#filtering-widget-data not working because `use Filament\Pages\Dashboard\Concerns\HasFiltersForm;` no exist.

**Comment by danielreales7** [comment]
@danharrin The same thing happens to me with ChartWidget.

If I use `protected static ?string $pollingInterval = 'keep-alive';` It is always sending requests to the server, which is wrong. 
Widgets should update based on filters.

I can show my code:

`class SalesChart extends ChartWidget
{
    use InteractsWithPageFilters;

    protected static ?string $heading = 'Total ventas por mes';
    protected static ?string $maxHeight = '275px';
    protected static ?string $pollingInterval = 'keep-alive';

    protected function getData(): array
    {
        $startDate = $this->filters['startDate'] ?? null;
        $endDate = $this->filters['endDate'] ?? null;
        $payment_method = $this->filters['payment_method'] ?? null;
        $store_id = Filament::getTenant()->id;

        $data = Trend::query(
            DailySale::when($payment_method, fn(Builder $query) => $query->where('payment_method', $payment_method))
                ->where('store_id', $store_id)
            )
            ->dateColumn('date')
            ->dateAlias('saleDate')
            ->between(
                start: Carbon::parse($startDate),
                end: Carbon::parse($endDate),
            )
            ->perMonth()
            ->sum('total');

        $transformedResults = $data->map(fn ($value) => new TrendValue(
            date: $value->date,
            aggregate: DailySale::castAttribute('total', $value->aggregate)
        ));

        return [
            'datasets' => [
                [
                    'label' => 'Total €',
                    'data' => $transformedResults->map(fn (TrendValue $value) => $value->aggregate),
                ],
            ],
            'labels' => $data->map(fn (TrendValue $value) => $value->date),
        ];
    }

    protected function getType(): string
    {
        return 'bar';
    }
}`

If I set it to null, it doesn't update the char

**Comment by danharrin** [comment]
Please open an issue with a reproduction repository

**Comment by danielreales7** [comment]
> Please open an issue with a reproduction repository

Done! [#10026](https://github.com/filamentphp/filament/issues/10026)

**Comment by danharrin** [comment]
Fixed by #10041.

**Pull request merged**